### PR TITLE
[Feature] Add BYOT Distillation

### PR DESCRIPTION
**Pull request by spynccat**
## Motivation

Add self-teacher method [BeYourOwnTeacher](https://github.com/luanyunteng/pytorch-be-your-own-teacher).

## Modification

1. Add `SelfDistill` distillation.
2. Add `BYOTDistiller` 
3. Add `KDSoftCELoss` as soft label loss in BYOT.
4. Add `BYOTConnector`.
5. Modify `L2Loss`, add arg `dist` to implement `torch.dist` function.
6. Add `connector_idx=None` `ConfigurableDistiller.get_record()` 
7. Add cifar100 autoaug config.
8. Add UT for `SelfDistill`, `BYOTDistiller`, `KDSoftCELoss`,  `BYOTConnector`.

### CODE REVIEW

File: mmrazor/models/algorithms/distill/configurable/self_distill.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from torch import nn
+
+from mmrazor.models.utils import add_prefix
+from mmrazor.registry import MODELS
+from ...base import BaseAlgorithm, LossResults
+
+
+@MODELS.register_module()
+class SelfDistill(BaseAlgorithm):
+    """``SingleTeacherDistill`` can be used to develop distill algorithms
```
Comment (HIT-cwh) [inline mmrazor/models/algorithms/distill/configurable/self_distill.py:15]: Pls fix typos.

File: mmrazor/models/algorithms/distill/configurable/self_distill.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        # The student loss will not be updated into ``losses`` in some
+        # pretraining process.
+        self.calculate_student_loss = calculate_student_loss
+
+        # In ``ConfigurableDistller``, the recorder manager is just
+        # constructed, but not really initialized yet.
+        self.distiller.prepare_from_student(self.student)
+        # Still prepare from self-teacher. Teacher recorders of
+        # ``SelfDistiller`` hook from self.student but require detach().
+        self.distiller.prepare_from_teacher(self.student)
```
Comment (HIT-cwh) [inline mmrazor/models/algorithms/distill/configurable/self_distill.py:48]: It seems that line 45 and line 48 are the same.

File: configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            depth=18,
+            num_stages=4,
+            out_indices=(3, ),
+            style='pytorch'),
+        neck=dict(type='mmcls.GlobalAveragePooling'),
+        head=dict(
+            type='mmcls.LinearClsHead',
+            num_classes=100,
+            in_channels=512,
+            loss=dict(type='mmcls.CrossEntropyLoss', loss_weight=1.0))),
```
Comment (HIT-cwh) [inline configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py:36]: Could we directly use `architecture=dict(cfg_path='mmcls::resnet/resnet18_8xb16_cifar10.py', pretrained=False)`

File: mmrazor/models/losses/l2_loss.py [hunk 4]
```diff
@@ -49,10 +53,14 @@ def forward(
...
-            loss = loss / s_feature.numel()
+        # Calculate l2_loss as dist.
+        if self.dist:
+            loss = torch.sqrt(loss)
         else:
-            loss = loss / s_feature.size(0)
+            if self.div_element:
+                loss = loss / s_feature.numel()
+            else:
+                loss = loss / s_feature.size(0)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/l2_loss.py:63]: For line67-75, whether the mean should be subtracted during normalization. If so, can we replace it with `F.layer_norm()`

File: configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            gt_labels=dict(type='ModuleInputs', source='head.loss_module')),
+        distill_losses=dict(
+            loss_fet_1=dict(
+                type='L2Loss', normalize=False, loss_weight=0.03, dist=True),
+            loss_label_1=dict(type='mmcls.CrossEntropyLoss', loss_weight=0.7),
+            loss_softl_1=dict(type='KDSoftCELoss', tau=3, loss_weight=0.3),
+            loss_fet_2=dict(
+                type='L2Loss', normalize=False, loss_weight=0.03, dist=True),
+            loss_label_2=dict(type='mmcls.CrossEntropyLoss', loss_weight=0.7),
+            loss_softl_2=dict(type='KDSoftCELoss', tau=3, loss_weight=0.3),
```
Comment (HIT-cwh) [inline configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py:55]: Can we use the `KLDivergence`  directly here?

File: configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py (hunk 3 above)
Comment (spynccat) [reply configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py:36]: MMCLS doesn't provide a `resnet18_cifar100` in `mmcls::resnet`. The model config dict is not loaded until `build_MODEL`, which means I cannot modify `architecture.head.num_classes` in config.py as in `mmcls:resnet.resnet50_8xb16_cifar100.py`.
An alternative option is to add a local config `mmrazor/configs/_base_/models/mmcls/resnet/resnet18_8xb16_cifar100.py`, but I'm not sure whether this is allowed.

File: mmrazor/models/algorithms/distill/configurable/self_distill.py (hunk 2 above)
Comment (spynccat) [reply mmrazor/models/algorithms/distill/configurable/self_distill.py:48]: Line 45 prepare `student_recorders` on student model with `prepare_from_student()`
Line 48 prepare `teacher_recorders` on student model with `prepare_from_teacher()`
In `SelfDistiller`, `teacher_recorders` record self-teacher vars on student model and require detach() in calculation.

File: mmrazor/models/algorithms/distill/configurable/self_distill.py (hunk 1 above)
Comment (spynccat) [reply mmrazor/models/algorithms/distill/configurable/self_distill.py]: fixed.

File: mmrazor/models/algorithms/distill/configurable/self_distill.py (hunk 2 above)
Comment (HIT-cwh) [reply mmrazor/models/algorithms/distill/configurable/self_distill.py:48]: My bad!

File: configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
+_base_ = [
```
Comment (sunnyxiaohu) [inline configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py:1]: what's the purpose of `_in1k.py` and `_in1k_kl.py`?

File: mmrazor/models/distillers/configurable_distiller.py [hunk 7]
```diff
@@ -237,7 +237,7 @@ def _check_loss_forward_mappings(
             loss_module = losses[loss_name]
             loss_forward_keys = signature(
                 loss_module.forward).parameters.keys()
-            assert len(loss_forward_keys) == len(forward_mappings.keys())
+            # assert len(loss_forward_keys) == len(forward_mappings.keys())
```
Comment (sunnyxiaohu) [inline mmrazor/models/distillers/configurable_distiller.py:240]: parameters without default value have to be checked.

Comment (spynccat) [reply mmrazor/models/distillers/configurable_distiller.py]: fixed.

File: configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py (hunk 6 above)
Comment (spynccat) [reply configs/distill/mmcls/byot/byot_logits_resnet18_cifar100_8xb16_in1k.py:1]: In gml config the soft label loss is KDsoftCE, while in paper KL.
Experiment shows a similar accuracy (top-1:80.63->80.66, top-5:95.57->95.76).
Now in `_in1k.py` KL is applied.

File: mmrazor/models/losses/l2_loss.py (hunk 4 above)
Comment (spynccat) [reply mmrazor/models/losses/l2_loss.py:63]: In gml source code mean is not subtracted, and only conduct on channel.
